feat: Support for login and registration via a browser custom tab

xitij2000 opened the pull request:
This change adds support for logging in and registering a new account using the
browser. This can be useful for cases where the only way to log into the
instatance is via a custom third-party auth provider.

openedx-webhooks commented:
Thanks for the pull request, @xitij2000!

### What's next?

*Please work through the following steps to get your changes ready for engineering review:*

#### :radio_button: Get product approval

If you haven't already, [check this list](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Does-my-contribution-require-Product-Review%3F) to see if your contribution needs to go through the product review process.

- If it does, you'll need to submit a product proposal for your contribution, and have it reviewed by the [Product Working Group](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3449028609/Product+Working+Group).
    - This process (including the steps you'll need to take) is documented [here](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Product-Review-Process).
- If it doesn't, simply proceed with the next step.

#### :radio_button: Provide context

To help your reviewers and other members of the community understand the purpose and larger context of your changes, feel free to add as much of the following information to the PR description as you can:

- Dependencies
  > This PR must be merged before / after / at the same time as ...
- Blockers
  > This PR is waiting for OEP-1234 to be accepted.
- Timeline information
  > This PR must be merged by XX date **because** ...
- Partner information
  > This is for a course on edx.org.
- Supporting documentation
- Relevant [Open edX discussion forum](https://discuss.openedx.org/) threads





#### :radio_button: Get a green build

If one or more checks are failing, continue working on your changes until this is no longer the case and your build turns green.


#### :radio_button: Update the status of your PR

Your PR is currently marked as a draft.  After completing the steps above, update its status by clicking "Ready for Review", or removing "WIP" from the title, as appropriate.

<!-- comment:end_of_wip -->


#### :radio_button: Let us know that your PR is ready for review:

### Who will review my changes?


This repository is currently maintained by `@openedx/openedx-mobile-maintainers`.  Tag them in a comment and let them know that your changes are ready for review.


### Where can I find more information?

If you'd like to get more details on all aspects of the review process for open source pull requests (OSPRs), check out the following resources:

- [Overview of Review Process for Community Contributions](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/FAQ-about-pull-requests.html)
- [Pull Request Status Guide](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/pull-request-statuses.html)
- [Making changes to your pull request](https://docs.openedx.org/en/latest/documentors/how-tos/make_changes_to_your_pull_request.html)

### When can I expect my changes to be merged?

Our goal is to get community contributions seen and reviewed as efficiently as possible.

However, the amount of time that it takes to review and merge a PR can vary significantly based on factors such as:

- The size and impact of the changes that it introduces
- The need for product review
- Maintenance status of the parent repository

:bulb: *As a result it may take up to several weeks or months to complete a review and merge your PR.*

<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IHRydWV9 -->

navinkarkera reviewed (approved):
@xitij2000 Nice work! :+1: 

:+1:
- [x] I tested this: (tested login in both real hardware as well as emulator)
- [x] I read through the code
- [x] I checked for accessibility issues
- [x] Includes documentation

mphilbrick211 commented:
@openedx/openedx-mobile-maintainers - is someone able to please take a look at this?

volodymyr-chekyrta commented:
> @openedx/openedx-mobile-maintainers - is someone able to please take a look at this?

I'll review it 👀 
@xitij2000, could you change the target branch to develop and fix conflicts if they appear?
Thank you!

xitij2000 commented:
> > @openedx/openedx-mobile-maintainers - is someone able to please take a look at this?
> 
> I'll review it 👀 @xitij2000, could you change the target branch to develop and fix conflicts if they appear? Thank you!

@volodymyr-chekyrta  I've changed the target branch; however, I've had some issues running the develop branch before due to it often being reliant on unmerged PRs in edx-platform. So I might need your help with that. 

Could you have a quick look and see if the general direction of the PR is okay or if bigger changes are needed so I can budget the time accordingly?

volodymyr-chekyrta commented:
> > > @openedx/openedx-mobile-maintainers - is someone able to please take a look at this?
> > 
> > 
> > I'll review it 👀 @xitij2000, could you change the target branch to develop and fix conflicts if they appear? Thank you!
> 
> @volodymyr-chekyrta I've changed the target branch; however, I've had some issues running the develop branch before due to it often being reliant on unmerged PRs in edx-platform. So I might need your help with that.
> 
> Could you have a quick look and see if the general direction of the PR is okay or if bigger changes are needed so I can budget the time accordingly?

@xitij2000 I checked the approach, and it is good 👍; I have a few comments according to the code, but it will be better to address them after resolving conflicts

xitij2000 commented:
> > > > @openedx/openedx-mobile-maintainers - is someone able to please take a look at this?
> > > 
> > > 
> > > I'll review it 👀 @xitij2000, could you change the target branch to develop and fix conflicts if they appear? Thank you!
> > 
> > 
> > @volodymyr-chekyrta I've changed the target branch; however, I've had some issues running the develop branch before due to it often being reliant on unmerged PRs in edx-platform. So I might need your help with that.
> > Could you have a quick look and see if the general direction of the PR is okay or if bigger changes are needed so I can budget the time accordingly?
> 
> @xitij2000 I checked the approach, and it is good 👍; I have a few comments according to the code, but it will be better to address them after resolving conflicts

I've rebased the code, but still need to test that everything works.

volodymyr-chekyrta commented on app/src/main/java/org/openedx/app/AppActivity.kt:
```diff
@@ -141,10 +150,15 @@ class AppActivity : AppCompatActivity(), InsetHolder, WindowSizeHolder {
                         addFragment(LogistrationFragment())
                     } else {
-                        addFragment(SignInFragment())
+                        val bundle = Bundle()
+                        bundle.putString("auth_code", authCode)
+                        val fragment = SignInFragment()
+                        fragment.arguments = bundle
+                        addFragment(fragment)
```
Please use `SignInFragment.newIntance(...)` method

volodymyr-chekyrta commented on auth/src/main/java/org/openedx/auth/data/repository/AuthRepository.kt:
```diff
@@ -43,6 +43,15 @@ class AuthRepository(
+            grantType = ApiConstants.GRANT_TYPE_CODE,
+            clientId = config.getOAuthClientId(),
+            code = code,
+            redirectUri = "${config.getApplicationID()}://oauth2Callback"
```
`oauth2Callback` to constants

volodymyr-chekyrta commented on auth/src/main/java/org/openedx/auth/presentation/logistration/LogistrationFragment.kt:
```diff
@@ -67,10 +68,23 @@ class LogistrationFragment : Fragment() {
+                            UrlUtils.openInBrowser(
+                                activity = context,
+                                apiHostUrl = viewModel.apiHostUrl,
+                                url = "/register",
```
Please move `/register` to constants

volodymyr-chekyrta commented on auth/src/main/java/org/openedx/auth/presentation/signin/SignInFragment.kt:
```diff
@@ -43,6 +43,11 @@ class SignInFragment : Fragment() {
                 val appUpgradeEvent by viewModel.appUpgradeEvent.observeAsState(null)
 
                 if (appUpgradeEvent == null) {
+                    val authCode = arguments?.getString("auth_code")
+                    if (authCode is String && !state.loginFailure && !state.loginSuccess) {
+                        arguments?.remove("auth_code")
```
Please move arguments like `"auth_code"` to companion object constants.
e.g.
```
companion object {
        private const val ARG_COURSE_ID = "courseId"
        private const val ARG_INFO_TYPE = "info_type"
        private const val ARG_AUTH_CODE = "auth_code"

        fun newInstance(courseId: String?, infoType: String?): SignInFragment {
            val fragment = SignInFragment()
            fragment.arguments = bundleOf(
                ARG_COURSE_ID to courseId,
                ARG_INFO_TYPE to infoType
            )
            return fragment
        }
    }
```

volodymyr-chekyrta commented on auth/src/main/java/org/openedx/auth/presentation/signin/SignInViewModel.kt:
```diff
@@ -158,11 +163,41 @@ class SignInViewModel(
+                    onUnknownError()
+                    _uiState.update { it.copy(loginFailure = true) }
+                }.onSuccess {
+                    logger.d { "Browser login success" }
```
I think we can remove the debug log from here

volodymyr-chekyrta commented on auth/src/main/java/org/openedx/auth/presentation/signin/compose/SignInView.kt:
```diff
@@ -298,13 +302,18 @@ private fun AuthForm(
                 textColor = MaterialTheme.appColors.primaryButtonText,
                 backgroundColor = MaterialTheme.appColors.secondaryButtonBackground,
                 onClick = {
+                    if(state.isBrowserLoginEnabled) {
+                        onEvent(AuthEvent.SignInBrowser)
```
`if(state.isBrowserLoginEnabled)` -> `if (state.isBrowserLoginEnabled)`

volodymyr-chekyrta commented on auth/src/main/java/org/openedx/auth/presentation/signin/compose/SignInView.kt:
```diff
@@ -298,13 +302,18 @@ private fun AuthForm(
                 onClick = {
+                    if(state.isBrowserLoginEnabled) {
+                        onEvent(AuthEvent.SignInBrowser)
+                    } else {
                     keyboardController?.hide()
```
```suggestion
                    } else {
                        keyboardController?.hide()
```

volodymyr-chekyrta commented on auth/src/main/java/org/openedx/auth/presentation/sso/BrowserAuthHelper.kt:
```diff
@@ -0,0 +1,32 @@
+    suspend fun signIn(activityContext: Activity) {
+        logger.d { "Browser-based auth initiated" }
+        val uri = Uri.parse("${config.getApiHostURL()}/oauth2/authorize").buildUpon()
+            .appendQueryParameter("client_id", config.getOAuthClientId())
+            .appendQueryParameter("redirect_uri", "${activityContext.packageName}://oauth2Callback")
+            .appendQueryParameter("response_type", "code").build()
```
It probably will be good to move the params to constants as well

volodymyr-chekyrta commented on core/src/main/java/org/openedx/core/config/Config.kt:
```diff
@@ -30,6 +30,10 @@ class Config(context: Context) {
         return getString(URI_SCHEME)
     }
 
+    fun getApplicationID(): String {
+        return getString(APPLICATION_ID, "")
+    }
+
```
duplicates:
```
fun getAppId(): String {
    return getString(APPLICATION_ID, "")
}
```

volodymyr-chekyrta commented on docs/how-tos/auth-using-browser.rst:
```diff
@@ -0,0 +1,45 @@
+How to user Browser-based Login and Registration
+================================================
+
```
Just wanted to point out that I like this doc 👍

volodymyr-chekyrta reviewed (changes requested):
Minor changes, mostly move value to constant:

xitij2000 commented:
@volodymyr-chekyrta  I will allocate some time soon to go through these properly and update the PR. Thanks for the great feedback!

xitij2000 commented:
@volodymyr-chekyrta  I've updated the PR based on your feedback. Do have a look!

volodymyr-chekyrta commented:
The code looks good to me.
@marcotuts, would you like to check this feature from a product perspective before merging?

xitij2000 commented on default_config/dev/config.yaml:
```diff
@@ -77,6 +77,10 @@ WHATS_NEW_ENABLED: false
 #feature flag to enable registration from app
 REGISTRATION_ENABLED: true
+#feature flag to do the authentication flow in the browser to log in
+BROWSER_LOGIN: true
```
Yes, of course! Sorry.

marcotuts commented:
Would it be possible to clarify the usage for this, and if possible any screenshots of it, even if just from local dev? 

I want to make sure I understand the flow here, is the configuration such that you are either using the native login / register (on by default?) or usign the browser mechanism? 

I ask becasue we also merged in a secondary web login button that can be optionally configured on the native view to direct users to a browser based login (ex: for corporate logins or other options not available natively). 

I still think the ability to choose a full browser login is a nimportant addition to the app, so I dont consider the two described features above as conflicting but I want to make sure I understand the changes and what the level of theming might be, or how the browser window renders (overlaps, embedded, etc). 

Thanks @xitij2000 for this contribution!

xitij2000 commented:
@marcotuts Sure! It will be hard to do it through screenshots since it a flow, but I will try to make a screencap and post it here. 

In this particular implementation, the web-based login repalces the native login. So when you click on the login button, it opens a custom browser tab, which is essentially a full-screen browser window (AFAIK it will use your default browser) where the OAuth2 based login flow will be initiated. When the user completes the log in in this browser windows (it could be using a username and password or whatever third-party auth system the LMS is configured to use) it will redirect back to the app with the code and the app will fetch a login token from that. 

From what I understand the feature you're describing adds the above login flow as an option during runtime rather than build time (which is what this PR does). I think if that is the case, we could simplify this PR to use that flow. Is my understanding correct?

xitij2000 commented:
@marcotuts  Here is a video of this at work:
[Screen_recording_20241209_172251.webm](https://github.com/user-attachments/assets/ce988634-5dcd-42c2-bd11-3ac1957c9a81)

xitij2000 commented:
> Thanks for the video! This looks great, thanks again for the contribution. It might be worth eventually exploring the embedded mode for this view but it seamlessly loads and returns to the app in the demo so that's 99% of what we need :D
> 
> product 👍


I am not very knowledgeable about Android so I followed the advice I heard about WebViews for OAuth2 which instead [recommended using custom tabs](https://support.google.com/faqs/answer/12284343?hl=en-GB).
